@here/harp-datasource-protocol/lib/ExprEvaluator.ts

```diff
@@ -220,6 +327,119 @@ export class ExprEvaluator implements ExprVisitor<Value, ExprEvaluatorContext> {
+                })
+            );
+        } else {
+            const input = context.evaluate(expr.input);
+
+            if (typeof input !== "number") {
+                throw new Error(`input '${input}' must be a number`);
+            }
+
+            if (input < expr.stops[0][0]) {
+                return context.evaluate(expr.defaultValue);
+            }
+
+            let index = expr.stops.findIndex(s => s[0] > input);
+
+            if (index === -1) {
+                index = expr.stops.length;
+            }
+
+            return context.evaluate(expr.stops[index - 1][1]);
+        }
+    }
+
+    visitInterpolateExpr(expr: InterpolateExpr, context: ExprEvaluatorContext): Value {
+        if (context.scope === ExprScope.Value) {
+            const input = context.evaluate(expr.input);
+            return new InterpolateExpr(
+                expr.mode,
+                Expr.isExpr(input) ? input : LiteralExpr.fromValue(input),
+                expr.stops.map(([key, value]) => {
+                    const v = context.evaluate(value);
+                    return [key, Expr.isExpr(v) ? v : LiteralExpr.fromValue(v)];
+                })
+            );
+        } else {
+            const param = context.evaluate(expr.input);
+
+            if (typeof param !== "number") {
+                throw new Error(`input must be a number`);
+            }
+
+            if (expr.mode[0] === "cubic") {
+                return cubicInterpolate(context, expr, param);
+            }
+
+            const keyIndex = expr.stops.findIndex(stop => stop[0] > param);
+
+            if (keyIndex === -1) {
+                // all the keys are smaller than the parameter
+                return context.evaluate(expr.stops[expr.stops.length - 1][1]);
+            } else if (keyIndex === 0) {
+                return context.evaluate(expr.stops[0][1]);
+            }
+
+            const [key, value] = expr.stops[keyIndex];
+            const [prevKey, prevValue] = expr.stops[keyIndex - 1];
+
+            const v0 = promoteValue(context.evaluate(prevValue));
+
+            let t = 0;
+
+            switch (expr.mode[0]) {
+                case "discrete":
+                    return v0;
+
+                case "linear":
+                    t = (param - prevKey) / (key - prevKey);
+                    break;
+
+                case "exponential": {
+                    t = Math.pow((param - prevKey) / (key - prevKey), expr.mode[1]);
+                    break;
+                }
+
+                default:
+                    throw new Error(
+                        `interpolation mode ${JSON.stringify(expr.mode)} is not supported`
+                    );
+            }
+
+            const v1 = promoteValue(context.evaluate(value));
+
+            if (typeof v0 === "number" && typeof v1 === "number") {
+                return THREE.MathUtils.lerp(v0, v1, t);
+            } else if (v0 instanceof Pixels && v1 instanceof Pixels) {
```

Comment:
For record - previous implementation supported "mix" of plain number with Pixels. of course it was buggy and i'm not sure if there is still use case.

Reply:
true, i think we should deprecate pixel literals and use world-ppi-scale, then the following will work
```
["interpolate", ...
  15, ["world-ppi-scale", 10],
  16, 1
```